diff --git a/include/wayland/EGL/eglplatform.h b/include/wayland/EGL/eglplatform.h
index 6d0d8db..81a5795 100644
--- a/include/wayland/EGL/eglplatform.h
+++ b/include/wayland/EGL/eglplatform.h
@@ -112,6 +112,14 @@ typedef intptr_t EGLNativeWindowType;
 
 #elif defined(__unix__) || defined(USE_X11)
 
+#if defined(MESA_EGL_NO_X11_HEADERS)
+
+typedef void            *EGLNativeDisplayType;
+typedef khronos_uintptr_t EGLNativePixmapType;
+typedef khronos_uintptr_t EGLNativeWindowType;
+
+#else
+
 /* X11 (tentative)  */
 #include <X11/Xlib.h>
 #include <X11/Xutil.h>
@@ -120,6 +128,8 @@ typedef Display *EGLNativeDisplayType;
 typedef Pixmap   EGLNativePixmapType;
 typedef Window   EGLNativeWindowType;
 
+#endif /* MESA_EGL_NO_X11_HEADERS */
+
 #elif defined(__APPLE__)
 
 typedef int   EGLNativeDisplayType;
